### ISSUE TITLE
Switching between breakpoints causes useBreakpoint to throw React state error 

### ISSUE THREAD

**Issue by tannerbaum** [chakra-ui/chakra-ui#2188]
# Bug report

## Describe the bug

When slowly hovering between breakpoints – or perhaps switching back and forth at the border – a React maximum update depth error is thrown from useBreakpoint. This is on RC5

## To reproduce

1. Go to a site using useBreakpoint
2. Find any of the breakpoints on the page 
3. Slowly hover around that breakpoint, possibly switching back and forth

## Minimal reproduction

Don't have one at the moment, could maybe produce one tomorrow 

## Expected behavior

Switches breakpoints without crashing the site 

## Screenshots

<img width="712" alt="Screenshot 2020-10-07 at 15 26 07" src="https://user-images.githubusercontent.com/1280597/95355849-4d65dc00-08c6-11eb-90d7-6c07ffeddc2d.png">

## System information

- OS: Mac OS Catalina 
- Browser (if applies): Chrome and Safari, seemingly not Firefox 
- Version of @chakra-ui/core: 1.0.0-rc.5
- Version of Node.js: 12.18.3
## Additional context

We had patched Chakra to remove the eventListener code breaking IE and Safari which is to be released in the next RC. The bug seems to occur with or without that change. Any advice would be great, since we for some reason are set to release a site and we are using release candidates in our dependencies 😅

**Comment by with-heart** [comment]
If you could provide a reproduction, that would be awesome and would certainly speed up the investigation a bit. :)

**Comment by tannerbaum** [comment]
@with-heart I tried last night to reproduce it in CodeSandbox and didn't have luck, but I have gotten closer to tracking it down and maybe this is enough to reveal the issue. 

We were trying to override the breakpoints from the theme and export it along with other custom theme properties in a useTheme hook. Here is the code:
```js
import {
  extendTheme,
  useTheme as useFullTheme,
  theme as defaultTheme,
} from "@chakra-ui/core";


//! This is the code that causes useBreakpoint to error
export const breakpoints = {
  sm: "600px",
  md: "800px",
  lg: "1140px",
  xl: "1140px",
};

// ...

const customTheme = {
  space,
  sizes,
  breakpoints,
  // ...
};

export const fullTheme = extendTheme({
  ...customTheme,
  fonts: {
    heading: customTheme.fonts.default,
    body: customTheme.fonts.default,
    ...customTheme.fonts,
  },
});

/** ChakraUI’s `useTheme` with typings for our theme */
export const useTheme = useFullTheme as () => typeof customTheme;
```

Here is the result logging out `theme.breakpoints` from Chakra directly and our useTheme hook's `breakpoints`:
<img width="373" alt="Screenshot 2020-10-09 at 10 21 41" src="https://user-images.githubusercontent.com/1280597/95561969-c6c31300-0a1b-11eb-9e3d-18a583211075.png">

With that known, what happens in Chakra is that when you hover on one of our override breakpoints, `current` and `breakpoint.breakpoint` rapidly alternate between the two breakpoint aliases on either side, i.e `sm, md / md, sm / sm, md / ....`

```js
  // use-breakpoint.ts
  const update = React.useCallback(
    (query: MediaQueryList, breakpoint: Breakpoint) => {
      if (query.matches && current !== breakpoint.breakpoint) {
        setCurrentBreakpoint(breakpoint)
      }
    },
    [current],
  )
```

**Comment by tannerbaum** [comment]
I think passing pixel values is the real problem here. Seems like without doing that the error doesn't occur. I also experimented with the different ways to set the breakpoints. It seems like the best way is definitely the `createBreakpoints` export which I only discovered in the source code 😅 

The last thing I will say is that my console.log I have been using inside the use-breakpoint snippet above doesn't seem to log the `md` breakpoint when using `createBreakpoints` for some reason? But the breakpoints appear to be triggered in the pages?
```js
const breakpoints = createBreakpoints({
  sm: "37.5em",
  md: "50em",
  lg: "71.25em",
  xl: "71.25em",
});
``` 

Anyway I hope something useful came out of all of this and I would love to hear feedback from the Chakra team if anything we are doing here is noticeably against the intentions of the library!

**Comment by ljosberinn** [comment]
`createBreakpoints` is a fairly recent addition and mentioned [here](https://next.chakra-ui.com/docs/migration#7-update-theme-breakpoints) in the docs :)

> The last thing I will say is that my console.log I have been using inside the use-breakpoint snippet above doesn't seem to log the md breakpoint when using createBreakpoints (and ems not px) for some reason?

Afraid I'll need a sandbox here.

We might look into throttling the listener and for a way to normalize units too as Chakras default theme uses `em`. Techincally that wouldn't matter though; it just merges in `base` at `0em` which is needed internally.
I didn't expect issues using `px` and I thought I added a note that mixing units is a conceptionally bad idea bu shouldn't cause issues but couldn't find it right now.

**Comment by kajchang** [comment]
Getting the same error still while using `em`.
Logging `console.log(theme.breakpoints, useBreakpoint(), window.innerWidth);`, it looks like it's cycling between `xl` and `3` when the width of the screen is exactly on the breakpoint. 

![Screen Shot 2020-10-13 at 11 02 39 AM](https://user-images.githubusercontent.com/39450215/95898316-b674a600-0d43-11eb-9cf4-b89eca77119c.png)

I can try to setup a reproduction later.

**Comment by tannerbaum** [comment]
@ljosberinn Servus! 

I think px values causing the problem makes some kind of sense when you look at how the media queries are generated.
https://github.com/chakra-ui/chakra-ui/blob/bfc91def1320f94dc93ba3452754f5fea74dfb3e/packages/media-query/src/create-media-query.ts#L53 
https://github.com/chakra-ui/chakra-ui/blob/bfc91def1320f94dc93ba3452754f5fea74dfb3e/packages/media-query/src/create-media-query.ts#L65-L67

If you put in a px value like 800px, that means the mediaQuery generated would be `(minWidth: /* previous breakpoint */ ) and (maxWidth: 799.99px)`. That would be a very fine line separating the breakpoints and I suppose if you hovered there like we did it would jump back and forth before exceeding the maximum depth of the setState calls. 

And one other point on that calculateMeasurement library you guys are using: it doesn't seem to work when the set breakpoints are decimal values. I did make a sandbox to illustrate that. The regex it uses doesn't cover that (very basic) usecase. So in the sandbox I included a potential replacement for the package that @hpohlmeyer wrote while we were looking at the breakpoints in our code 😄 
https://codesandbox.io/s/chakra-calculate-measurement-hso0u?file=/index.html
